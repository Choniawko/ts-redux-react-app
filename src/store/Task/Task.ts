export interface Task {
  id: number
  title: string
  description: string
}

export interface TaskState {
  tasks: ReadonlyArray<Task>
  task: Task | null
  count: number
  previos: number | null
  next: number | null
  error: string
}
