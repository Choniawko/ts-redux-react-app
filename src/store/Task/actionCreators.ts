import { Task } from "."
import ActionCreator from "../actionCreator"

export default {
  getTasks: ActionCreator<null>("TASK_GET_TASKS"),
  updateTasks: ActionCreator<Task[]>("TASK_UPDATE_TASKS"),
  getTask: ActionCreator<number>("TASK_GET_TASK"),
  updateTask: ActionCreator<Task>("TASK_UPDATE_TASK"),
  setError: ActionCreator<string>("TASK_SET_ERROR"),
}
