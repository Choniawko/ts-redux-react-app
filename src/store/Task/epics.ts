import { combineEpics, Epic, ofType } from "redux-observable"
import { of } from "rxjs"
import { catchError, mergeMap } from "rxjs/operators"
import { actionCreators } from "."
import { Action, RootState } from ".."
import { ajax, apiUrl } from "../../common/services"

const { getTasks, getTask, updateTasks, updateTask, setError } = actionCreators

const getTasksEpic: Epic<Action<null>, Action<any>, RootState> = actions$ =>
  actions$.pipe(
    ofType(getTasks.type),
    mergeMap(() =>
      ajax.get(apiUrl("tasks")).pipe(
        mergeMap(({ response }) => of(updateTasks.create(response))),
        catchError(error => of(setError.create(error))),
      ),
    ),
  )
const getTaskEpic: Epic<Action<number>, Action<any>, RootState> = actions$ =>
  actions$.pipe(
    ofType(getTask.type),
    mergeMap(id =>
      ajax.get(apiUrl(`task/${id}`)).pipe(
        mergeMap(({ response }) => of(updateTask.create(response))),
        catchError(error => of(setError.create(error))),
      ),
    ),
  )
export const epics = combineEpics(getTasksEpic, getTaskEpic)
