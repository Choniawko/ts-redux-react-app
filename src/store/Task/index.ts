export * from "./Task"
export { default as actionCreators } from "./actionCreators"
export { default as taskReducer, INITIAL_STATE } from "./reducer/reducer"
export { epics as taskEpics } from "./epics"
