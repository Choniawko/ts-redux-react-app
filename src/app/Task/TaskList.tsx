import React, { useEffect } from "react"
import { connect } from "react-redux"
import { RouteComponentProps } from "react-router-dom"
import { RootState } from "../../store"
import { actionCreators, Task } from "../../store/Task"

type ParamsProps = null
export interface Props extends RouteComponentProps<ParamsProps> {
  tasks: ReadonlyArray<Task>
  getTasks: () => void
}

const TaskList = ({ tasks, getTasks }: Props): JSX.Element => {
  useEffect(() => {
    getTasks()
  }, [])
  return (
    <>
      <h3>TaskList</h3>
      <div className="list">
        {tasks.map(({ id, title }) => (
          <div key={id}>{title}</div>
        ))}
      </div>
    </>
  )
}
const mapStateToProps = ({ task: { tasks } }: RootState) => ({
  tasks,
})
const mapDispatchToProps = {
  getTasks: actionCreators.getTasks.create,
}
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TaskList)
