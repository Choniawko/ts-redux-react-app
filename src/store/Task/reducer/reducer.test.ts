import { actionCreators, INITIAL_STATE, taskReducer } from ".."
import { testTask } from "../../../fixtures/Task"

describe("taskReducer", () => {
  it("calls updateTasks action", () => {
    const action = {
      type: actionCreators.updateTasks.type,
      payload: [testTask],
    }
    expect(taskReducer(INITIAL_STATE, action)).toMatchSnapshot()
  })
  it("calls updateTask action", () => {
    const action = {
      type: actionCreators.updateTask.type,
      payload: testTask,
    }
    expect(taskReducer(INITIAL_STATE, action)).toMatchSnapshot()
  })
})
