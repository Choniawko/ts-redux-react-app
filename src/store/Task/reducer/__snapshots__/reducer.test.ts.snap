// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`taskReducer calls updateTask action 1`] = `
Object {
  "count": 0,
  "error": "",
  "next": null,
  "previos": null,
  "task": Object {
    "description": "description",
    "id": 1,
    "title": "Title",
  },
  "tasks": Array [],
}
`;

exports[`taskReducer calls updateTasks action 1`] = `
Object {
  "count": 0,
  "error": "",
  "next": null,
  "previos": null,
  "task": null,
  "tasks": Array [
    Object {
      "description": "description",
      "id": 1,
      "title": "Title",
    },
  ],
}
`;
